crates/provider/src/new.rs

```diff
@@ -91,13 +155,223 @@ impl<N, T> RootProviderInner<N, T> {
+    /// such should not be overridden.
+    #[inline]
+    async fn watch_pending_transaction(
+        &self,
+        config: PendingTransactionConfig,
+    ) -> TransportResult<PendingTransaction> {
+        self.root().watch_pending_transaction(config).await
+    }
+
+    /// Subscribe to new block headers.
+    ///
+    /// Returns a stream of new block headers.
+    #[cfg(feature = "pubsub")]
+    async fn subscribe_blocks(&self) -> TransportResult<Subscription<Block>> {
+        self.root().pubsub_frontend()?;
+        let id = self.client().prepare("eth_subscribe", ("newHeads",)).await?;
+        self.root().get_subscription(id).await
+    }
+
+    /// Subscribe to new pending transactions.
+    ///
+    /// Returns a stream of new pending transaction hashes.
+    #[cfg(feature = "pubsub")]
+    async fn subscribe_pending_transactions(&self) -> TransportResult<Subscription<B256>> {
+        self.root().pubsub_frontend()?;
+        let id = self.client().prepare("eth_subscribe", ("newPendingTransactions",)).await?;
+        self.root().get_subscription(id).await
+    }
+
+    /// Subscribe to new full pending transactions.
+    ///
+    /// Returns a stream of new pending transactions.
+    #[cfg(feature = "pubsub")]
+    async fn subscribe_full_pending_transactions(
+        &self,
+    ) -> TransportResult<Subscription<N::TransactionResponse>> {
+        self.root().pubsub_frontend()?;
+        let id = self.client().prepare("eth_subscribe", ("newPendingTransactions", true)).await?;
+        self.root().get_subscription(id).await
+    }
+
+    /// Subscribe to an RPC event.
+    #[cfg(feature = "pubsub")]
+    #[auto_impl(keep_default_for(&, &mut, Rc, Arc, Box))]
+    async fn subscribe<P, R>(&self, params: P) -> TransportResult<Subscription<R>>
+    where
+        P: RpcParam,
+        R: RpcReturn,
+        Self: Sized,
+    {
+        self.root().pubsub_frontend()?;
+        let id = self.client().prepare("eth_subscribe", params).await?;
+        self.root().get_subscription(id).await
+    }
+
+    /// Cancels a subscription given the subscription ID.
+    #[cfg(feature = "pubsub")]
+    async fn unsubscribe(&self, id: U256) -> TransportResult<()> {
+        self.root().unsubscribe(id)
+    }
+
+    /// Watch for new blocks by polling the provider with
+    /// [`eth_getFilterChanges`](Self::get_filter_changes).
+    ///
+    /// Returns a builder that is used to configure the poller. See [`PollerBuilder`] for more
+    /// details.
+    ///
+    /// # Examples
+    ///
+    /// Get the next 5 blocks:
+    ///
+    /// ```no_run
+    /// # async fn example<N: alloy_network::Network>(provider: impl alloy_provider::Provider<N>) -> Result<(), Box<dyn std::error::Error>> {
+    /// use futures::StreamExt;
+    ///
+    /// let poller = provider.watch_blocks().await?;
+    /// let mut stream = poller.into_stream().flat_map(futures::stream::iter).take(5);
+    /// while let Some(block_hash) = stream.next().await {
```

Comment:
Can we get rid of the `into_stream().flat_map(futures::stream::iter)`? When do we want to use the stream vs the iter? https://docs.rs/futures/latest/futures/stream/fn.iter.html